### ISSUE TITLE
Support TLS out of the box

### ISSUE THREAD

**Issue by davewasmer** [denali-js/denali#96]
Denali apps should include TLS by default, if possible. We could use self-signed certs in dev mode, or potentially bake in a first class integration with LetsEncrypt.

**Comment by knownasilya** [comment]
👍  for LetsEncrypt integration.

**Comment by knownasilya** [comment]
Maybe even https://github.com/davewasmer/marathon-dns integration?

**Comment by davewasmer** [comment]
Some recent thoughts:

Most production node deployments I've seen do SSL/TLS termination at an nginx or load balancing layer in front of node, not inside node itself. If that's generally true, I'm not sure I see the value of having TLS enabled in development. Seems like it adds complexity, further diverges dev and prod, and doesn't really gain anything.

I'm curious to know what others have seen - is SSL/TLS termination done in node in any production apps you've seen?

cc @knownasilya @tomdale @acorncom

**Comment by tomdale** [comment]
For me, the importance of supporting TLS during development is so your local JavaScript app can talk to it over a secure origin. You're right that diverging dev and prod is generally a bad idea, but in this case, TLS should be more or less an abstraction underneath HTTP, whereas the browser's security model is often affected by whether the origin is TLS-enabled or not. Browser vendors are using new features as a carrot to push developers towards TLS (like Service Worker, and even retroactively for App Cache) so it's not clear in the future what things will work differently if it's not TLS end-to-end.

**Comment by knownasilya** [comment]
We've done TLS in node before and it worked fine, but it also depends on your deployment strategy. If you deploy to the cloud in GCP you set a cert in their infrastructure and forget about it, in an alpha instance on EC2 we had nginx and used that for certs, so there are many ways.

I do agree with Tom though, especially because of new features requiring SSL, like geolocation and service workers, it has become more of a pain to develop locally without SSL and no one wants to setup nginx with letsencrypt locally, it's a pain.

**Comment by acorncom** [comment]
Seems like it could be a plugin that ran in development and optionally was available to be turned on in production if needed.  Heroku is another example of where all SSL terminates above the app ...

**Comment by davewasmer** [comment]
Makes sense, wasn't aware of that trend in browsers (thanks for the context @tomdale). Let's :shipit:

**Comment by davewasmer** [comment]
Alright, I've been doing some research here. A bit of a braindump, sorry for the noise:

### TLS in production

I like the idea of a Let's Encrypt integration here. Ideally you could manually supply your certificates, or opt into the Let's Encrypt certs.

As I understand it, there's three parts to managing certs via ACME/Let's Encrypt:

1. Account authorization - ACME specifies that the client should generate a keypair, and once that client successfully proves domain ownership, that keypair is granted domain management rights.
2. Domain validation - the ACME server can issue a challenge to the client to demonstrate it's ownership of the domain it wants certs for. Typically this is "put a file at this random URL" or "put a DNS TXT record at this random subdomain".
3. Cert issuance - the ACME server actually sends the certificates once domain ownership has been proven.

Denali needs to support everything from basic, single process deployments on persistent hardware, all the way through horizontally scaled, multi-instance, clustered deployments on ephemeral hardware. I'm thinking that means the account keypair and certs need to be managed as Denali Models, saved to whatever data store the user sets up.

As for domain validation, we can't support DNS out of the box (who knows how your DNS is managed), so it's really just the "put a file here" challenge. Denali could add a route behind the scenes with a built in Action that responds with the correct challenge response.

For cert issuance & renewal, we would need to store the certs centrally (see above). We also would need to have clustering implemented, so we could do a rolling restart the HTTP servers when the certs renew to avoid any downtime.

#### Implementation

Looks like https://github.com/Daplie/node-letsencrypt is the standard right now for Node ACME clients. It requires two types of plugins to work: a certificate store and a challenge handler. Looks like we'll need to build custom Denali plugins for both types.

### TLS in development

Let's Encrypt won't work here, since the ACME server won't be able to hit our "localhost" domain to verify it.

That means we need self-signed certificates. It seems straightforward to generate them - the problem is that because they are self-signed, not signed by a trusted CA, the OS won't immediately trust the certs, so you'll get an annoying "Untrusted cert" warning every time.

On macOS at least, we could shell out to a command to trust the self-signed cert, but it requires sudo. So we could either (a) run with sudo every time, or (b) fail on startup if you are trying to run with SSL but we haven't trusted the certs yet. I'd lean towards (b). I'm not sure if you can elevate privileges of the Node process after it starts ...

There's also the question of managing those self-signed certs. Do we generate a unique cert per project and store it locally? That certainly feels cleaner, but then old projects leave a bunch of trusted yet obsolete certs in your local machine's trust store. The other option is to generate a single Denali cert for all Denali apps. It creates "magical" global state, but eliminates the cluttered trust store, and the "run sudo the first time for each Denali app".

@knownasilya you mentioned using marathon-dns, and other tutorials have mentioned setting up a local DNS forwarder to resolve a TLD like `.dev`. I think I'm missing something - why is that required here? Couldn't we just create self-signed certs for localhost?

**Comment by acorncom** [comment]
Thanks for the run down @davewasmer.  Based on what you're seeing there, my suggestion would be that we punt this past the v0.1.0 milestone.  At least for my uses, I'd rather see built-in support in Denali land for mailers, jobs (and possibly auth / websocket support as well) before being concerned about this.  Thoughts?

**Comment by davewasmer** [comment]
I'd agree that other things are higher priority right now, from a feature completeness perspective. Especially since you can run Denali with SSL right now (at least theoretically). My only concern is that this is a potentially good "marketing" feature. Some of the ideas that differentiate Denali are a little tough to explain at first, but saying "SSL in dev out of the box" is simple and compelling.

So yea, I agree, let's punt for now, but I'd like to avoid this falling through the cracks.

**Comment by acorncom** [comment]
Sounds good. And agreed, SSL in dev out of the box sounds slick ;-)

**Comment by davewasmer** [comment]
Random note: we could potentially use something like https://www.npmjs.com/package/configstore to manage a system-wide denali root ca or something like that - makes managing that "magic global state" from above a bit easier.

**Comment by davewasmer** [comment]
More braindump here after more research.

I think the most elegant solution here for development SSL is to have Denali generate a single root CA cert, saved system-wide, and generate certs on-the-fly for each app. I _think_ that should do it.

The trick then is to get browsers to trust that root CA cert. Most browsers delegate to the underlying OS store. Looks like Windows and Mac have relatively straightforward command line tools to add a cert to that store, so on first-run of your first local SSL app, we would ask for sudo and shell out to those commands to add the cert to the store.

There's one edge case: Firefox. It insists on shipping with it's own trust store, and ignores the OS. Plus, the command line tools to add certs to that store require separate, different installations on every OS. I don't love the idea of install additional system-wide software just to add the cert to Firefox. However - if you attempt to open a cert in Firefox, if offers a nice GUI to add it to the store. So I'm thinking we might just print out a notice on that first-run with a `file://` url and say "If you use Firefox, open this URL"

The other tricky part is domain names. Normal CAs won't issue certs for localhost because apparently some systems (Mac included) will attempt to resolve `localhost` via network DNS servers first. I'd like to follow their footsteps as well, plus automatic `my-app.dev` URLs is just a nice feature to have.

For Mac and Linux this is relatively straightforward - use `/etc/resolver/*` or `/etc/resolv.conf` respectively. The OS will try those DNS servers first, and we can register our own there which will intercept *.dev requests and point to the local machine.

For Windows - this is apparently impossible. It doesn't seem to have any concept of using a DNS server for only certain domains. The only option seems to be editing the Windows equivalent of the `/etc/hosts` file every time, which would require elevated permissions every time. We could either edit it each time you run `denali server`, or we could edit it the first time and leave it to the user to clean up at some point.

## Conclusions

* For Firefox users on any OS, they will have to click a link and follow the Firefox GUI to install a root cert the first time they use Denali
* The Windows story kinda sucks  😞

**Comment by davewasmer** [comment]
A big step closer now: https://github.com/davewasmer/devcert
